File: BOJ/DP/2579.cpp
// 2579 계단 오르기
// https://www.acmicpc.net/problem/2579
#include <iostream>
#include <vector>
#define LL long long

using namespace std;

LL mx(LL a, LL b) {
    return a > b ? a : b;
}

LL solution(int N, vector<int> steps) {
    LL answer = 0;
    vector<LL> dp(N, 0);

    dp[0] = steps[0];
    dp[1] = steps[0] + steps[1];
    dp[2] = mx(steps[0] + steps[2], steps[1]+steps[2]);

    for (int i = 3; i < N; ++i)
    {
        // 
        dp[i] = mx(dp[i-3]+steps[i]+steps[i-1], dp[i-2] + steps[i]); 
    }

    //cout << dp[N-1] << '\n';
    answer = dp[N-1];
    return answer;
}

int main(void) {
    int N;
    vector<int> steps;

    cin >> N;
    steps = vector<int> (N, 0);

    for (int i = 0; i < N; ++i)
    {
        cin >> steps[i];
    }

    cout << solution(N, steps);
    return 0;
}

File: BOJ/DP/2579.py
# 2579 계단 오르기
# https://www.acmicpc.net/problem/2579
import sys

input = sys.stdin.readline


def solution(n, steps):
    dp = [0 for _ in range(300)]

    dp[0] = steps[0]
    dp[1] = steps[0]+steps[1]
    dp[2] = max(steps[1]+steps[2], steps[0]+steps[2])

    for i in range(3, n):
        dp[i] = max(dp[i-3]+steps[i-1]+steps[i], dp[i-2]+steps[i])

    return dp[n-1]


n = int(input())
steps = [0 for _ in range(300)]

for i in range(n):
    steps[i] = int(input())


print(solution(n, steps))
